libraries/Microsoft.Bot.Builder.Azure/CosmosDbPartitionedStorage.cs

```diff
@@ -0,0 +1,256 @@
+                    RealId = change.Key,
+                    Document = json,
+                };
+
+                var etag = (change.Value as IStoreItem)?.ETag;
+                if (etag == null || etag == "*")
+                {
+                    // if new item or * then insert or replace unconditionally
+                    await _container.UpsertItemAsync(
+                            documentChange,
+                            new PartitionKey(documentChange.PartitionKey),
+                            cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else if (etag.Length > 0)
+                {
+                    // if we have an etag, do opt. concurrency replace
+                    await _container.UpsertItemAsync(
+                            documentChange,
+                            new PartitionKey(documentChange.PartitionKey),
+                            new ItemRequestOptions() { IfMatchEtag = etag, },
+                            cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new Exception("etag empty");
+                }
+            }
+        }
+
+        public async Task DeleteAsync(string[] keys, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            foreach (var key in keys)
+            {
+                await _container.DeleteItemAsync<DocumentStoreItem>(
+                    partitionKey: new PartitionKey(key),
+                    id: CosmosDbKeyEscape.EscapeKey(key),
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Connects to the CosmosDB database and creates / gets the container.
+        /// </summary>
+        private async Task InitializeAsync()
+        {
+            if (_container == null)
+            {
+                var cosmosClientOptions = _cosmosDbStorageOptions.CosmosClientOptions ?? new CosmosClientOptions();
+
+                _client = new CosmosClient(
+                    _cosmosDbStorageOptions.CosmosDbEndpoint,
+                    _cosmosDbStorageOptions.AuthKey,
+                    cosmosClientOptions);
+
+                _container = await _client
+                        .GetDatabase(_cosmosDbStorageOptions.DatabaseId)
+                        .CreateContainerIfNotExistsAsync(
+                            _cosmosDbStorageOptions.ContainerId,
+                            DocumentStoreItem.PartitionKeyPath,
+                            _cosmosDbStorageOptions.ContainerThroughput)
+                        .ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Internal data structure for storing items in a CosmosDB Collection.
+        /// </summary>
+        private class DocumentStoreItem
```

Comment:
Needs to extend IStoreItem or Read fails

Reply:
Done. Thanks and good catch. (Odd as it was before - not sure how it reverted :) )